File: pysrc/fractal/mind/glimpse/_op/interpret.py
"""
Interpret
=========

Runs a single Monte-Carlo tree search simulation, updating
all traversed placements.

"""

from math import sqrt

from ..glimpsing import Glimpsing


class Interpret(
    Glimpsing,
):
    def interpret(
        self,
        flow: Flowing,
        eta: float,
        root: bool = True,
    ) -> Qvalue:
        state = flowing.state()
        actions = flowing.actions()
        info, new_node = self.analyze(state)

        if new_node:
            return info.Vest
        else:
            if root:
                epsilon = self.epsilon_noise
            else:
                epsilon = 0

        scores = self.upper_confidence_bounds(
            info,
            self.cpuct,
            epsilon,
            eta,
        )
        action_id = argmax(scores)
        action = actions[action_id]

        flow.flow(placement)
        wr = flow.white_energy()
        if flow.at_dawn():
            r = wr
        else:
            r = -wr

        shall_cortex_change = wr != flow.at_dawn()

        qnext = self.interpret(
            flow,
            eta,
            root=False,
        )

        if shall_cortex_change:
            qnext = -qnext

        qvalue = r + self.eta * qnext

        self.update(
            state,
            action_id,
            qvalue,
        )
        self.total_nodes_traversed += 1
        return

File: src/fractal/demon/glimpse/_op/interpret.rs
//! Update the state of the glimpse with information.

pub trait Interpret {
    fn interpret(&self, realized: Realized, action: Action, energy: Energy) {};
}

impl Interpret for Glimpsing {
    fn interpret(
        &self,
        realized: WorldState,
        action: Action,
        energy: Energy,
    ):
        stats = worldstate.tree<state>.stats
        astats = stats<action_id>
        stats<action_id> = PlacementAnalysis(
            astats.P,
            astats.W + q,
            astats.N + 1,
        )
}
